File: 2018-04-26/W3D4/W3D4/ViewController.m
//
//  ViewController.m
//  W3D4
//
//  Created by Jason Liang on 4/26/18.
//  Copyright © 2018 jasonliang. All rights reserved.
//

#import "ViewController.h"

@interface ViewController () <UITableViewDataSource>

@property (nonatomic, strong) NSArray *results;
@property (nonatomic, strong) UITableView *tableView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView = [[UITableView alloc] initWithFrame:self.view.bounds];
    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"cell"];
    [self.view addSubview:self.tableView];
    self.tableView.dataSource = self;
    
    NSString *urlString = @"https://swapi.co/api/people";
    NSURL *url = [NSURL URLWithString:urlString];
    NSURLRequest *request;
//    NSMutableURLRequest
    request = [[NSURLRequest alloc] initWithURL:url];
//    request.HTTPMethod = @"POST";
//    NSData *data = [@"some sort body" dataUsingEncoding:NSUTF8StringEncoding];
//    request.HTTPBody = data;
    NSURLSessionTask *task;
    NSURLSession *session = [NSURLSession sharedSession];
    task = [session dataTaskWithRequest:request
                      completionHandler:
            ^(NSData * _Nullable data,
              NSURLResponse * _Nullable response,
              NSError * _Nullable error) {
                NSLog(@"got result");
                if (error) {
                    //do something
                }
                [self parseResponseData:data];
    }];
    [task resume];
//    [task cancel];

}

- (void)parseResponseData:(NSData *)data {
    NSError *error;
    NSDictionary *result = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
    if (error) {
        //do something
    }
    self.results = result[@"results"];
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.tableView reloadData];
    });
    
}

#pragma mark - UITableViewDataSource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.results.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    NSDictionary *item = self.results[indexPath.row];
    cell.textLabel.text = item[@"name"];
    return cell;
}


@end

File: 2017-11-21/W4D1/W4D1/ViewController.swift
//
//  ViewController.swift
//  W4D1
//
//  Created by Jason Liang on 11/21/17.
//  Copyright © 2017 Jason Liang. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDataSource {
  
  var stations : [[String:Any]] = []
  let tableView = UITableView()
  
  override func viewDidLoad() {
    super.viewDidLoad()
    self.view.addSubview(self.tableView)
    self.tableView.frame = self.view.bounds
    self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
    self.tableView.dataSource = self
    retrieveBikeStationInfo()
  }

  //MARK: UITableViewDataSource
  func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return self.stations.count
  }
  
  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: "cell")
    let station = self.stations[indexPath.row]
    if let name = station["name"] as? String {
      cell?.textLabel?.text = name
    }
    return (cell != nil) ? cell! : UITableViewCell()
  }
  
  //MARK:
  
  func retrieveBikeStationInfo() {
    let session = URLSession.shared
    
    let urlString = "https://tor.publicbikesystem.net/ube/gbfs/v1/en/station_information"
    
    guard let url = URL(string: urlString) else {
      return
    }
    
    let task = session.dataTask(with: url) { (data, response, error) in
      print("i got something something back!")
      guard let data = data else { return }
      guard let json = try? JSONSerialization.jsonObject(with: data, options: []) else { return }
      guard let stationInfoJson = json as? [String:Any] else { return }
      guard let stationsJson = stationInfoJson["data"] as? [String:Any] else { return }
      guard let stations = stationsJson["stations"] as? [[String:Any]] else { return }
      
      print(stations)
      self.stations = stations;
      
      DispatchQueue.main.async {
        self.tableView.reloadData()
      }
    }
    
    task.resume()
  }

}












